proposals/0252-callable.md

```diff
@@ -0,0 +1,610 @@
+# Introduce callables
+
+* Proposal: SE-NNNN
+* Authors: [Richard Wei](https://github.com/rxwei), [Dan Zheng](https://github.com/dan-zheng)
+* Review Manager: TBD
+* Status: Implementation in progress
+* Implementation: https://github.com/apple/swift/pull/23517
+
+## Introduction
+
+Note: throughout this document, let "call-syntax" refer to the syntax of applying a function to arguments: `f(x, y, z)`.
+
+This proposal introduces [callables](https://en.wikipedia.org/wiki/Callable_object) to Swift. Callables are values that define function-like behavior and can be applied using function application syntax.
+
+In a nutshell, we propose to introduce a new declaration syntax with the keyword `call` :
+
+```swift
+struct Adder {
+    var base: Int
+    call(_ x: Int) -> Int {
+        return base + x
+    }
+}
+```
+
+Values that have a `call` member can be applied like functions, forwarding arguments to the `call` member.
+
+```swift
+let add3 = Adder(base: 3)
+add3(10) // => 13
+```
+
+## Motivation
+
+Currently, in Swift, only a few kinds of values are syntactically callable:
+- Values with function types.
+- Type names (e.g. `T` can be called like `T(...)`, which is desugared to `T.init(...)`).
+- Values with a `@dynamicCallable` type.
+
+However, callable syntax can also be useful for other values, primarily those that behave like functions. This includes:
+- Values that represent functions: mathematical functions, function expressions, etc.
+- Values that have one main use and want to provide a simple callable interface: neural network layers, parsers, efficient bound closures, etc.
+
+Here are some concrete sources of motivation.
+
+### Values representing functions
+
+Values of some types exactly represent functions: in the mathematical sense (a mapping from inputs to outputs), or in the context of programming languages.

```

Comment:
`Values of some types exactly represent functions` is a bit confusing, particularly `some types`.
```suggestion
Some types exactly represent functions: in the mathematical sense (a mapping from inputs to outputs), or in the context of programming languages.
```

Reply:
"Types representing functions" is incorrect. I intentionally changed the original "some types ..." to "values of some types".

Functions are values, not types. Thus types cannot represent functions.

Reply:
How about `Some values exactly represent functions?`

Reply:
That is technically correct, but it is correct more because "some values" can just _be_ functions, which feels like tautology and doesn't go into more details than the section title.
I changed it to "Values of some nominal types exactly represent functions".